### PR TITLE
Left recursion, Left Factoring, Trie, Unit Tests

### PR DESCRIPTION
**Pull request by bisakhmondal**
# Left Recursion
## Description
- [x] Left Recursion Done
- [ ] Indirect left recursion (if required, after going through the grammar, a decision needs to be made).

### CODE REVIEW

**Comment by bisakhmondal** [comment]
WILL REBASE IT WHEN I'LL HAVE A GOOD MOOD 🐍

File: src/include/trie/memoryEfficientTrie.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
+#ifndef JUCC_MEMORYEFFICIENTTRIE_H
+#define JUCC_MEMORYEFFICIENTTRIE_H
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
```
Comment (noob77777) [inline src/include/trie/memoryEfficientTrie.h]: prefer moving this file to utils.

File: src/include/trie/memoryEfficientTrie.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#include "grammar/grammar.h"
+
+namespace jucc {
+
```
Comment (noob77777) [inline src/include/trie/memoryEfficientTrie.h]: namespace util

File: src/trie/memoryEfficientTrie.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
+#include "trie/memoryEfficientTrie.h"
+
+jucc::Trie::Trie() { count = 0; }
+
+jucc::TrieManager::TrieManager() { master = NewTrieNode(); }
+
+void jucc::TrieManager::Insert(const grammar::Entity &entities) {
+  auto *head = master;
+  grammar::Entity tillKeys;
```
Comment (noob77777) [inline src/trie/memoryEfficientTrie.cpp]: variable naming till_keys
readability-identifier-naming

File: src/include/trie/memoryEfficientTrie.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ public:
+  TrieManager();
+  [[nodiscard]] Trie *GetMaster() const;
+
+  void Insert(const grammar::Entity &);
+
+  void InsertAll(const grammar::Production &);
+
+  static void GreedyPreorder(Trie *, int &, grammar::Entity &, bool);
```
Comment (noob77777) [inline src/include/trie/memoryEfficientTrie.h]: Add comments to header files

File: src/include/utils/leftRecursion.h [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#ifndef JUCC_LEFTRECURSION_H
+#define JUCC_LEFTRECURSION_H
+#include "grammar/grammar.h"
+
+namespace jucc::utils {
+const char DASH[] = "'";
+
+grammar::Productions RemoveDirectLeftRecursion(const grammar::Production &);
```
Comment (noob77777) [inline src/include/utils/leftRecursion.h]: clang-tidy moment without variable name in function parameter

File: src/include/utils/leftRecursion.h [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#ifndef JUCC_LEFTRECURSION_H
+#define JUCC_LEFTRECURSION_H
```
Comment (noob77777) [inline src/include/utils/leftRecursion.h]: inappropriate header file name.
left_recursion.h

File: src/trie/memoryEfficientTrie.cpp [hunk 7]
```diff
@@ -0,0 +1,58 @@
+#include "trie/memoryEfficientTrie.h"
+
```
Comment (noob77777) [inline src/trie/memoryEfficientTrie.cpp]: inappropriate file name

File: src/include/utils/leftFactoring.h [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#ifndef JUCC_LEFTFACTORING_H
```
Comment (noob77777) [inline src/include/utils/leftFactoring.h]: in appropriate file name

File: src/include/trie/memoryEfficientTrie.h [hunk 9]
```diff
@@ -0,0 +1,43 @@
+#ifndef JUCC_MEMORYEFFICIENTTRIE_H
+#define JUCC_MEMORYEFFICIENTTRIE_H
```
Comment (noob77777) [inline src/include/trie/memoryEfficientTrie.h]: in appropriate file name
memory_efficient_trie.h

File: src/grammar/grammar.cpp [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  if (prefix.size() > entities.size()) {
+    return false;
+  }
+  for (int i = 0; i < static_cast<int>(prefix.size()); i++) {
+    if (entities[i] != prefix[i]) {
+      return false;
+    }
+  }
+  return true;
+}
```
Comment (TheSYNcoder) [inline src/grammar/grammar.cpp]: Can it be better like this 

namespace jucc {
    std::string grammar::Rule::toString() const {
             ....
    }

}

File: src/include/grammar/grammar.h [hunk 11]
```diff
@@ -0,0 +1,42 @@
+#ifndef JUCC_GRAMMAR_H
+#define JUCC_GRAMMAR_H
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace jucc {
+namespace grammar {
+const char EPSILON[] = "epsilon";
+typedef std::vector<std::string> Entity;
```
Comment (TheSYNcoder) [inline src/include/grammar/grammar.h]: modernize-use-using

Use of using.

File: src/include/grammar/grammar.h [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+  [[nodiscard]] const std::vector<std::string> &getEntities() const { return entities; }
+  void setEntities(const std::vector<std::string> &entities) { Rule::entities = entities; }
+  [[nodiscard]] std::string toString() const;
+  [[nodiscard]] bool HasPrefix(const Entity &) const;
+};
+
+typedef std::vector<grammar::Rule> Rules;
+class Production {
+  std::string parent;
+  std::vector<Rule> rules;
```
Comment (TheSYNcoder) [inline src/include/grammar/grammar.h]: Please name variables with suffix _.

File: src/include/grammar/grammar.h [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+ public:
+  Production() = default;
+  Production(std::string parent, std::vector<Rule> rules) : parent(std::move(parent)), rules(std::move(rules)) {}
+  [[nodiscard]] const std::string &getParent() const { return parent; }
+  [[nodiscard]] const std::vector<Rule> &getRules() const { return rules; }
+  void setParent(const std::string &parent) { Production::parent = parent; }
+  void setRules(const std::vector<Rule> &rules) { Production::rules = rules; }
+};
+
+typedef std::vector<Production> Productions;
```
Comment (TheSYNcoder) [inline src/include/grammar/grammar.h]: modernize-use-using.

File: src/include/trie/memoryEfficientTrie.h [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace jucc {
+
+class Trie {
+ public:
+  // std::vector<Trie *> children;
+  std::unordered_map<std::string, Trie *> nodes;
+  grammar::Entity keys_list;
+  int count;
+  Trie();
```
Comment (TheSYNcoder) [inline src/include/trie/memoryEfficientTrie.h]: Make the default constructor initialize count =0 .
Trie() : count(0) {}

File: src/include/grammar/grammar.h [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+ public:
+  Production() = default;
+  Production(std::string parent, std::vector<Rule> rules) : parent(std::move(parent)), rules(std::move(rules)) {}
+  [[nodiscard]] const std::string &getParent() const { return parent; }
+  [[nodiscard]] const std::vector<Rule> &getRules() const { return rules; }
+  void setParent(const std::string &parent) { Production::parent = parent; }
+  void setRules(const std::vector<Rule> &rules) { Production::rules = rules; }
+};
+
```
Comment (TheSYNcoder) [inline src/include/grammar/grammar.h]: Please write more comments.
Please make the codebase user-friendly.
Thanks.

**Review by noob77777 — CHANGES_REQUESTED** [review]

File: src/include/grammar/grammar.h (hunk 15 above)
Comment (noob77777) [reply src/include/grammar/grammar.h]: I have added comments on main rebase will fix this.

File: src/include/grammar/grammar.h (hunk 12 above)
Comment (noob77777) [reply src/include/grammar/grammar.h]: rebase will fix this

File: src/include/grammar/grammar.h (hunk 13 above)
Comment (noob77777) [reply src/include/grammar/grammar.h]: rebase will fix this

File: src/include/utils/leftRecursion.h (hunk 5 above)
Comment (bisakhmondal) [reply src/include/utils/leftRecursion.h]: epic moments

File: src/include/trie/memoryEfficientTrie.h (hunk 2 above)
Comment (bisakhmondal) [reply src/include/trie/memoryEfficientTrie.h]: make sense :+1:

File: src/trie/memoryEfficientTrie.cpp (hunk 3 above)
Comment (bisakhmondal) [reply src/trie/memoryEfficientTrie.cpp]: done

File: src/include/trie/memoryEfficientTrie.h (hunk 4 above)
Comment (bisakhmondal) [reply src/include/trie/memoryEfficientTrie.h]: Ohh god!!!

File: test/utils/utils_test.cpp [hunk 16]
```diff
@@ -0,0 +1,162 @@
+#include "gtest/gtest.h"
```
Comment (noob77777) [inline test/utils/utils_test.cpp:1]: left_recursion_test.cpp

File: test/utils/utils_test.cpp [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+  ASSERT_EQ(lf_removed[0].GetParent(), "E");
+
+  ASSERT_EQ(lf_removed[0].GetRules().size(), 3);
+
+  ASSERT_EQ(lf_removed[0].GetRules()[0].ToString(), "ieStSt");
+  ASSERT_EQ(lf_removed[0].GetRules()[1].ToString(), "a");
+  ASSERT_EQ(lf_removed[0].GetRules()[2].ToString(), "b");
+}
+
+TEST(jucc, LeftFactoringComplex) {
```
Comment (noob77777) [inline test/utils/utils_test.cpp]: wrong test suit

File: test/trie/trie_test.cpp [hunk 18]
```diff
@@ -0,0 +1,25 @@
+#include "grammar/grammar.h"
```
Comment (noob77777) [inline test/trie/trie_test.cpp]: test/utils/trie_test.cpp

File: test/trie/trie_test.cpp [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+#include "gtest/gtest.h"
+#include "trie/memoryEfficientTrie.h"
+namespace grammar = jucc::grammar;
+
+TEST(trie, TRIETest1) {
```
Comment (noob77777) [inline test/trie/trie_test.cpp]: trie1
using "test" in test name is unethical

File: test/utils/utils_test.cpp [hunk 20]
```diff
@@ -0,0 +1,162 @@
+#include "gtest/gtest.h"
+#include "utils/leftFactoring.h"
+#include "utils/leftRecursion.h"
+
+namespace grammar = jucc::grammar;
+
+TEST(utils, directLeftRecursion) {
```
Comment (noob77777) [inline test/utils/utils_test.cpp]: inconsistent test naming choose either camelCase or PascalCase

File: src/include/utils/left_factoring.h [hunk 21]
```diff
@@ -0,0 +1,13 @@
+#ifndef JUCC_LEFT_FACTORING_H
+#define JUCC_LEFT_FACTORING_H
+#include "grammar/grammar.h"
+namespace grammar = jucc::grammar;
+
+namespace jucc::utils {
+grammar::Productions RemoveLeftFactors(const grammar::Production & /*prod*/);
```
Comment (noob77777) [inline src/include/utils/left_factoring.h:11]: Header file function declarations not commented

File: src/include/utils/left_recursion.h [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+#define JUCC_LEFT_RECURSION_H
+#include "grammar/grammar.h"
+
+namespace jucc::utils {
+const char DASH[] = "'";
+
+grammar::Productions RemoveDirectLeftRecursion(const grammar::Production & /*prod*/);
+bool IsRecursive(const grammar::Production & /*prod*/);
+
+grammar::Productions RemoveIndirectLeftRecursions(const grammar::Productions &);
```
Comment (noob77777) [inline src/include/utils/left_recursion.h]: Function declaration unnamed parameter.
Not commented header file

File: src/include/utils/trie/memory_efficient_trie.h [hunk 23]
```diff
@@ -0,0 +1,80 @@
+#ifndef JUCC_UTILS_MEMORYEFFICIENTTRIE_H
+#define JUCC_UTILS_MEMORYEFFICIENTTRIE_H
+#include <sstream>
```
Comment (noob77777) [inline src/include/utils/trie/memory_efficient_trie.h:3]: header-guard should be JUCC_UTILS_MEMORY_EFFICIENT_TRIE_H

File: test/utils/utils_test.cpp [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+  ASSERT_EQ(rules_e_dash.size(), 2);
+
+  ASSERT_EQ(rules_e_dash[0].ToString(), "+TE" + std::string(jucc::utils::DASH));
+  ASSERT_EQ(rules_e_dash[1].ToString(), std::string(grammar::EPSILON));
+
+  ASSERT_EQ(rules_e[0].ToString(), "TE" + std::string(jucc::utils::DASH));
+  ASSERT_EQ(rules_e[1].ToString(), std::string(grammar::EPSILON) + "E" + std::string(jucc::utils::DASH));
+}
+
+TEST(jucc, MaxLenPrefix0) {
```
Comment (noob77777) [inline test/utils/utils_test.cpp]: Test suite name should be utils.

File: src/include/grammar/grammar.h [hunk 25]
```diff
@@ -17,15 +18,19 @@ class Rule {
...
  public:
   Rule() = default;
-  explicit Rule(std::vector<std::string> entities) : entities_(std::move(entities)) {}
-  [[nodiscard]] const std::vector<std::string> &GetEntities() const { return entities_; }
-  void SetEntities(const std::vector<std::string> &entities) { Rule::entities_ = entities; }
+  explicit Rule(Entity entities) : entities_(std::move(entities)) {}
+  [[nodiscard]] const Entity &GetEntities() const { return entities_; }
+  void SetEntities(const Entity &entities) { Rule::entities_ = entities; }
+  [[nodiscard]] std::string ToString() const;
+  [[nodiscard]] bool HasPrefix(const Entity & /*prefix*/) const;
```
Comment (noob77777) [inline src/include/grammar/grammar.h:35]: Comment HasPrefix method .

**Review by noob77777 — CHANGES_REQUESTED** [review]

File: src/include/utils/left_recursion.h (hunk 22 above)
Comment (bisakhmondal) [reply src/include/utils/left_recursion.h]: Just kind of an interface, no implementation required till now.

File: src/include/grammar/grammar.h (hunk 25 above)
Comment (bisakhmondal) [reply src/include/grammar/grammar.h:35]: Bro!! Self-describing.

File: src/include/utils/left_factoring.h (hunk 21 above)
Comment (bisakhmondal) [reply src/include/utils/left_factoring.h:11]: haaat bc

File: src/include/grammar/grammar.h (hunk 25 above)
Comment (noob77777) [reply src/include/grammar/grammar.h:35]: No

File: src/include/utils/left_factoring.h (hunk 21 above)
Comment (noob77777) [reply src/include/utils/left_factoring.h:11]: ektu likhe de trie.h ta toh aesthetic lagche

File: src/include/grammar/grammar.h (hunk 25 above)
Comment (bisakhmondal) [reply src/include/grammar/grammar.h:35]: Done!!

File: src/include/utils/left_factoring.h (hunk 21 above)
Comment (bisakhmondal) [reply src/include/utils/left_factoring.h:11]: Done!!

File: src/include/utils/left_recursion.h (hunk 22 above)
Comment (noob77777) [reply src/include/utils/left_recursion.h]: 👀 Not commented header file

Comment (bisakhmondal) [reply src/include/utils/left_recursion.h]: Done!!

**Comment by noob77777** [comment]
Looks great 👍  merging to main.

**Review by noob77777 — APPROVED** [review]

**Pull request merged**